state-sync/state-sync-v2/data-streaming-service/src/streaming_client.rs

```diff
@@ -0,0 +1,343 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    data_stream::{DataStreamListener, NotificationId},
+    error::Error,
+};
+use async_trait::async_trait;
+use diem_types::transaction::Version;
+use futures::{
+    channel::{mpsc, oneshot},
+    stream::FusedStream,
+    SinkExt, Stream,
+};
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+// A useful type definition for readability.
+pub type Epoch = u64;
+
+/// The streaming client used by state sync to fetch data from the Diem network
+/// to synchronize local state.
+///
+/// Note: the streaming service streams data sequentially, so clients (e.g.,
+/// state sync) can process data notifications in the order they're received.
+/// For example, if we're streaming transactions with proofs, state sync can
+/// assume the transactions are returned in monotonically increasing versions.
+#[async_trait]
+pub trait DataStreamingClient {
+    /// Fetches all account states at the specified version. The specified
+    /// version must be an epoch ending version, otherwise an error will be
+    /// returned. Account state proofs are at the same specified version.
+    async fn get_all_accounts(&self, version: Version) -> Result<DataStreamListener, Error>;
+
+    /// Fetches all epoch ending ledger infos starting at `start_epoch`
+    /// (inclusive) and ending at the last known epoch advertised in the network.
+    async fn get_all_epoch_ending_ledger_infos(
+        &self,
+        start_epoch: Epoch,
+    ) -> Result<DataStreamListener, Error>;
+
+    /// Fetches all transactions with proofs from `start_version` to
+    /// `end_version` (inclusive), where the proof versions can be up to the
+    /// specified `max_proof_version` (inclusive). If `include_events` is true,
+    /// events are also included in the proofs.
+    async fn get_all_transactions(
+        &self,
+        start_version: Version,
+        end_version: Version,
+        max_proof_version: Version,
+        include_events: bool,
+    ) -> Result<DataStreamListener, Error>;
+
+    /// Fetches all transaction outputs with proofs from `start_version` to
+    /// `end_version` (inclusive), where the proof versions can be up to the
+    /// specified `max_proof_version` (inclusive).
+    async fn get_all_transaction_outputs(
+        &self,
+        start_version: Version,
+        end_version: Version,
+        max_proof_version: Version,
+    ) -> Result<DataStreamListener, Error>;
+
+    /// Refetches the payload for the data notification corresponding to the
+    /// specified `notification_id`.
+    ///
+    /// Note: this is required because data payloads may be invalid, e.g., due
+    /// to invalid or malformed data returned by a misbehaving peer or a failure
+    /// to verify a proof. The refetch request forces a refetch of the payload
+    /// and the `refetch_reason` notifies the streaming service as to why the
+    /// payload must be refetched.
+    async fn refetch_notification_payload(
```

Comment:
I went for `refetch_notification_payload` because it emphasizes that we want the data in the notification and not the notification itself. In some edge cases, it's possible that the notification failed because the payload was too large (could not be requested in one notification), hence we need to fetch the payload, which may then result in multiple notifications (hence why this returns a stream). I'll leave this as is for now.